### ISSUE TITLE
About pWaitDstStageMask

### ISSUE THREAD

**Issue by lordalcol** [SaschaWillems/Vulkan#433]
Hi Sascha,

thank you for your examples, they are the most useful Vulkan samples that can be found.
By reading the spec and checking some of your code, my team have noticed something that might be not 100% correct. You seem to suggest that the `pWaitDstStageMask` determines what stage from a **previous** operation needs to be waited for before starting the render pass, as you say in the comment:

```
// Wait for color attachment output to finish before rendering the text overlay
VkPipelineStageFlags stageFlags = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
submitInfo.pWaitDstStageMask = &stageFlags;
```

Our understanding of the spec, that says (https://www.khronos.org/registry/vulkan/specs/1.0/man/html/VkSubmitInfo.html)

> pWaitDstStageMask is a pointer to an array of pipeline stages at which each corresponding semaphore wait will occur.

is that the `pWaitDstStageMask` is actually the stage of the *new render pass* that should be blocked by the wait semaphore.

In other words, for something like the text overlay, I think `pWaitDstStageMask` should be `VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT`, because that's the stage of the text overlay pass that needs to be blocked until the previous semaphore is signaled. (aka we cannot render text to the framebuffer until everything else has been drawn, but we can do vertex shading for example).

This probably applies to other places where `pWaitDstStageMask`  is used.

**Comment by cforfang** [comment]
Your description is right, so the comment seems wrong/misleading, but I don't believe  ``VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT`` is needed here since this stage doesn't write any pixels to the framebuffer. This only happens in ``VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT``.

Figuring out exactly what's needed here is a bit non-trivial because of the interactions between the semaphores between the submits, ``pWaitDstStageMask``, layout transitions inside subpasses, and ``VK_SUBPASS_EXTERNAL`` subpass-dependencies. ``pWaitDstStageMask`` by itself isn't enough -- you need a full dependency chain from the previous render-pass to the next (also accounting for any layout transitions which must be blocked).

**Comment by lordalcol** [comment]
You are right that at `VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT ` stage stuff is not written/blended to the framebuffer, good point. The comment is a bit misleading and it got me to do some other thing wrong in my code.

**Comment by SaschaWillems** [comment]
Thanks a lot for bringing this up and yes, I think you're right on this, the comment is misleading. Will take a look at this.

**Comment by cforfang** [comment]
For my own understanding's sake -- please correct me if I'm wrong! -- here's how things seems to work at the moment and why it seems correct. Apologies for the wall of text. :)

* In ``VulkanExampleBase::submitFrame()`` there's a branch for ``submitOverlay``. If this is _not_ taken ``vkQueuePresent`` is set up to wait on ``semaphores.renderComplete`` and then called. No issues here.
* If the branch is taken, ``UIOverlay->cmdBuffers[currentBuffer]`` is submitted and set up to wait on ``semaphore.renderComplete`` in ``pWaitDstStageMask = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT``.
*  This command buffer is set up to render to the swap-chain image that is going to be presented later. It's doing this using a render-pass with ``loadOp = LOAD``, i.e. it's rendering on top of the already prepared image. Additionally, it's transitioning the image from ``initialLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR`` to ``colorReference.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL`` during the render-pass then back to ``finalLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR`` once the render-pass is finished.
* For synchronization, there are subpass-dependencies from ``VK_SUBPASS_EXTERNAL`` to subpass 0 with ``srcStageMask = VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT`` and ``dstStageMask = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT``, as well as 0 -> ``VK_SUBPASS_EXTERNAL`` with a dependency from ``VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT`` to ``VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT``.
* The ``VK_SUBPASS_EXTERNAL`` -> 0 subpass-dependency guarantees that the image transition won't happen too early. Remember, the semaphore wait is in ``COLOR_ATTACHMENT_OUTPUT``, meaning the image might still be in use by previously submitted operations until we reach that stage and wait. Therefore we need the external dependency blocking the transition until the right moment (if we didn't it could happen immediately after submit -- long before we reach ``COLOR_ATTACHMENT_OUTPUT`` -- thus also before the image is finished being used by previous operations).
* Because layout transitions are specced to happens-after the ``EXTERNAL`` dependency's ``srcStageMask`` (here: ``BOTTOM_OF_PIPE_BIT``) and before ``dstStageMask`` (here: ``COLOR_ATTACHMENT``), we have a dependency-graph like this: [New rendering to image] BLOCKED ON [image-transition] (because ``dstStageMask = COLOR`` in external -> 0 subpass-dependency) BLOCKED ON [semaphore signaling (because ``srcStageMask = BOTTOM`` in external -> 0 subpass-dependency is after ``COLOR_ATTACHMENT_OUTPUT`` where the semaphore wait will occur). 
* When subpass 0 reaches ``COLOR_ATTACHMENT_OUTPUT`` the image is in ``COLOR_ATTACHMENT_OPTIMAL`` because the transition must've completed.
* Subpass completes, image is transitioned to ``LAYOUT_PRESENT_SRC_KHR``.
* During command-buffer submission, ``semaphores.overlayComplete`` is set to be signaled when the overlay-command-buffer is finished. ``vkQueuePresent`` is also set up to wait on this, so everything will have finished by the time present happens.

So everything seems correct, although I guess it would be clearer if the subpass-dependency's ``srcStageMask`` was ``COLOR_ATTACHMENT`` because then it's obvious the image transition is also blocked on the semaphore.

**Comment by lordalcol** [comment]
> This command buffer is set up to render to the swap-chain image that is going to be presented later. It's doing this using a render-pass with loadOp = LOAD, i.e. it's rendering on top of the already prepared image. Additionally, it's transitioning the image from initialLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR to colorReference.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL during the render-pass then back to finalLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR once the render-pass is finished.

In this part, are you saying that the src layout transition is basically defined by the combination of the `VkAttachmentDescription .initialLayout` and the `VkAttachmentReference .layout` values?

What does it mean that the `initialLayout` is set as  `VK_IMAGE_LAYOUT_UNDEFINED` in most examples I have found?

It appears as if the `VkAttachmentDescription .finalLayout` defines the final image transition, so it's confusing that the initial one is defined in the `VkAttachmentReference`.

Thanks.

**Comment by chrisvarns** [comment]
Please tell me if I'm incorrect:

- The subpass dependencies are defining at what stages, and for what operations during those stages, we must wait for the layout transitions to have occurred (the API doesn't make this clear to me, I guess that a layout transition is a member of the `MEMORY_READ_BIT` set of operations?).

- The synchronization could alternatively be specified on entry/exit to the subpass with vkCmdPipelineBarrier commands?

**Comment by cforfang** [comment]
> In this part, are you saying that the src layout transition is basically defined by the combination of the VkAttachmentDescription .initialLayout and the VkAttachmentReference .layout values?

Yes exactly. The first transition here is from the layout the image is in (``initialLayout``) into what we want it to be in subpass 0 (``layout``). Similarly for ``finalLayout``; the driver knows which layout the image is in during the last subpass (here ``COLOR_ATTACHMENT_OPTIMAL`` in subpass 0) and we tell it what we want it to end up as (here ``PRESENT_SRC_KHR``). All transitions are in pairs basically.

 If we had more subpasses there could be transitions between these as well, for example ``COLOR_ATTACHMENT_OPTIMAL`` in subpass 0 (where we write to an image) and ``SHADER_READ_ONLY_OPTIMAL`` in subpass 1 (where we read from the image). For correct operation there must be additional subpass-dependencies to make sure the driver does these transitions at the right time as well.

> What does it mean that the initialLayout is set as VK_IMAGE_LAYOUT_UNDEFINED in most examples I have found?

Doing a layout transition on an image isn't necessarily free; it might involve non-trivial GPU work to -- for example -- decompress the image from some framebuffer-compression format used during rendering into something that can be read by some other GPU-unit (or the other way around from something readable to something that's optimal for rendering into). If you _don't_ use ``VK_IMAGE_LAYOUT_UNDEFINED`` the image contents are required to be preserved by this conversion -- but often you don't care to preserve the image so doing this would be wasted work.

That's why there's ``UNDEFINED`` which is basically an optimization where you tell the driver _"I don't care what was in the image before (and what its layout was), just do whatever's quickest to get the image into this new layout so I can use it"_. 

As an example, instead of spending GPU-cycles re-compressing an image into some kind of compressed ``COLOR_ATTACHMENT_OPTIMAL`` format, the driver can do whatever is required to initialize the image in the new layout (which might just be writing a short header or similar). This if often what you want at the start of a new frame; you'll have some image in e.g. ``LAYOUT_SHADER_READ_ONLY_OPTIMAL`` from the last frame but want to get it into ``COLOR_ATTACHMENT_OPTIMAL`` so you can render to it. There's no point in the driver spending cycles transitioning the contents in the old layout to the new layout (since you'll immediately overwrite it anyway), so it's a lot more efficient to use ``UNDEFINED`` here.

(Note that even when you transition from ``UNDEFINED`` you still need the correct dependencies in place as you _are_ changing the image and these changes must happen at the correct  time.)

>The subpass dependencies are defining at what stages, and for what operations during those stages, we must wait for the layout transitions to have occurred

Half right, basically :) In addition they also defines what must occur _before_ the transitions can start (e.g. what the transition itself must wait on and which caches must be flushed first and so on), and similarly what operations must happen _after_ the transition to make the changes visible to the correct stages and operations that will use it later (e.g. the transition might change image data so the texture cache should be flushed to avoid reading stale data).

>  (the API doesn't make this clear to me, I guess that a layout transition is a member of the MEMORY_READ_BIT set of operations?).

I think the short answer is you don't know since it's up to the driver to figure out what a transition actually means on the current hardware. This is why what you actually specify in the subpass-dependencies (and pipeline-barriers for that matter) is what you've _done_ to the image that must be flushed ('made available') and then also _what you're going to use the image for_ and thus what operations need to have their cached invalidated (so that any changes to the image done by the transition is made 'visible' to these operations).

> The synchronization could alternatively be specified on entry/exit to the subpass with vkCmdPipelineBarrier commands?

There's some very particular rules about how ``vkCmdPipelineBarrier`` can be used inside render-passes which means the short answer is 'no', but if you modify the question to be about _renderpasses_ and ``vkCmdPipelineBarrier`` between these then yes (e.g. you can have a 1-subpass-render-passes with ``initialLayout == layout == finalLayout`` and do all transitions outside the render-pass using  ``vkCmdPipelineBarrier``. That's actually often easier since keeping track of the interactions between pipeline-barriers, semaphores, and subpass-dependencies (including ``VK_SUBPASS_EXTERNAL`` dependencies) can be quite tricky...

**Comment by chrisvarns** [comment]
Thanks, this is very helpful!

**Comment by lordalcol** [comment]
Agreed, thank you Christian
